libraries/platforms/dromajo-vcs/bsg_manycore_platform.cpp

```diff
@@ -0,0 +1,518 @@
+// (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+// LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+// ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+#include <bsg_manycore_platform.h>
+#include <bsg_manycore.h>
+#include <bsg_manycore_printing.h>
+#include <bsg_manycore_config.h>
+
+#include <bp_hb_platform.h>
+
+#include <set>
+
+/* these are convenience macros that are only good for one line prints */
+#define manycore_pr_dbg(mc, fmt, ...) \
+        bsg_pr_dbg("%s: " fmt, mc->name, ##__VA_ARGS__)
+
+#define manycore_pr_err(mc, fmt, ...) \
+        bsg_pr_err("%s: " fmt, mc->name, ##__VA_ARGS__)
+
+#define manycore_pr_warn(mc, fmt, ...) \
+        bsg_pr_warn("%s: " fmt, mc->name, ##__VA_ARGS__)
+
+#define manycore_pr_info(mc, fmt, ...) \
+        bsg_pr_info("%s: " fmt, mc->name, ##__VA_ARGS__)
+
+/********************************** BlackParrot platform API **********************************/
+
+/*
+ * Reads the manycore bridge for number of credits used in the endpoint
+ * @param[in] credits_used --> Pointer to a location in memory that will hold the number of credits used
+ * @returns HB_MC_SUCCESS
+ */
+int bp_hb_get_credits_used(int *credits_used) {
+	uint32_t *bp_to_mc_req_credits_addr = (uint32_t *) (MC_BASE_ADDR + BP_TO_MC_REQ_CREDITS_ADDR);
+	*credits_used = (int) *bp_to_mc_req_credits_addr;
+	if (*credits_used < 0) {
+		bsg_pr_err("Credits used cannot be negative. Credits used = %d", *credits_used);
+		return HB_MC_FAIL;
+	}
+	return HB_MC_SUCCESS;
+}
+
+/*
+ * Writes a 128-bit manycore packet in 32-bit chunks to the manycore bridge FIFO
+ * @param[in] pkt --> Pointer to the manycore packet
+ * @returns HB_MC_SUCCESS
+ * TODO: Implement error checking (Requires some modifications in Dromajo)
+ */
+int bp_hb_write_to_mc_bridge(hb_mc_packet_t *pkt) {
+	uint32_t *bp_to_mc_req_fifo_addr = (uint32_t *) (MC_BASE_ADDR + BP_TO_MC_REQ_FIFO_ADDR);
+	for(int i = 0; i < 4; i++) {
+		*bp_to_mc_req_fifo_addr = pkt->words[i];
+		bp_to_mc_req_fifo_addr++;
+	}
+	return HB_MC_SUCCESS;
+}
+
+/*
+ * Checks if the MC to BP FIFO contains any valid elements to be read
+ * @param[in] entries --> Pointer to a location in memory that will hold the number of entries
+ * @param[in] type --> Type of FIFO to read from
+ * @returns HB_MC_SUCCESS on success or HB_MC_FAIL on fail
+ */
+int bp_hb_get_fifo_entries(int *entries, hb_mc_fifo_rx_t type) {
+	switch (type) {
+		case HB_MC_FIFO_RX_REQ:
+		{
+			uint32_t *mc_to_bp_req_fifo_entries_addr = (uint32_t *) (MC_BASE_ADDR + MC_TO_BP_REQ_ENTRIES_ADDR);
+			*entries = *mc_to_bp_req_fifo_entries_addr;
+			if (*entries < 0) {
+				bsg_pr_err("Entries occupied cannot be negative. Entries = %d", *entries);
+				return HB_MC_FAIL;
+			}
```

Comment:
Would refactor this and keep the address calculation inside the case statement. Move the mechanics outside, because it's duplicated code